File: Problems/Algorithms/278. First Bad Version/FirstBadVersion.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */

class Solution: VersionControl() {
    override fun firstBadVersion(n: Int) : Int {
        var left = 1
        var right = n
        
        while (left < right) {
            val mid = left + (right - left) / 2
            val isBad = isBadVersion(mid)
            
            if (isBad) {
                right = mid
            } else {
                left = mid + 1
            }
        }
        
        return right
	}
}

File: Problems/Algorithms/278. First Bad Version/first_bad_version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        left, right = 1, n

        ans = -1
        while left <= right:
            mid = left + (right - left) // 2
            is_bad = isBadVersion(mid)

            if is_bad:
                ans = mid
                right = mid - 1
            else:
                left = mid + 1

        return ans

